presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/WeightedDoubleReservoirSample.java

```diff
@@ -0,0 +1,214 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.reservoirsample;
+
+import io.airlift.slice.SizeOf;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.SliceOutput;
+import io.airlift.slice.Slices;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.Arrays;
+import java.util.concurrent.ThreadLocalRandom;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class WeightedDoubleReservoirSample
+        implements Cloneable
+{
+    public static final int MAX_SAMPLES_LIMIT = 1_000_000;
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(WeightedDoubleReservoirSample.class).instanceSize();
+
+    private int count;
+    private double[] samples;
+    private double[] weights;
+
+    public WeightedDoubleReservoirSample(int maxSamples)
+    {
+        checkArgument(maxSamples > 0, "Maximum number of samples must be positive");
+        checkArgument(maxSamples <= MAX_SAMPLES_LIMIT, "Maximum number of samples must not exceed " + MAX_SAMPLES_LIMIT);
+
+        this.samples = new double[maxSamples];
+        this.weights = new double[maxSamples];
+    }
+
+    private WeightedDoubleReservoirSample(WeightedDoubleReservoirSample other)
+    {
+        this.count = other.count;
+        this.samples = Arrays.copyOf(other.samples, other.samples.length);
+        this.weights = Arrays.copyOf(other.weights, other.weights.length);
+    }
+
+    private WeightedDoubleReservoirSample(int count, double[] samples, double[] weights)
+    {
+        this.count = count;
+        this.samples = requireNonNull(samples, "samples is null");
+        this.weights = requireNonNull(weights, "weights is null");
+    }
+
+    public long getMaxSamples()
+    {
+        return samples.length;
+    }
+
+    public void add(double sample, double weight)
+    {
+        checkArgument(weight >= 0, format("Weight %s cannot be negative", weight));
+        double adjustedWeight = Math.pow(
+                ThreadLocalRandom.current().nextDouble(),
+                1.0 / weight);
+        addWithAdjustedWeight(sample, adjustedWeight);
+    }
+
+    private void addWithAdjustedWeight(double sample, double adjustedWeight)
+    {
+        if (count < samples.length) {
+            samples[count] = sample;
+            weights[count] = adjustedWeight;
+            count++;
+            bubbleUp();
+            checkArguments();
```

Comment:
this call doesn't belong here (arguments must be checked in the very beginning of the method); remove it